File: percolation/percolation.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sat Apr 13 16:30:18 2019

@author: musahni
"""

import numpy as np
import matplotlib.pyplot as plt

class UnionFind(object):
    def __init__(self, size):
        self.size = size
        self.id = list(range(size))
        self.depth = [1]*self.size
        

    def find(self, p):
        while p != self.id[p]:
            p = self.id[p]
        
        return p
    
    def connected(self, p, q):
        return self.find(p) == self.find(q)
    
    def union(self, p, q):
        rootp = self.find(p)
        rootq = self.find(q)
        
        if rootp == rootq:
            pass
        
        if self.depth[rootp] < self.depth[rootq]:
            self.id[rootp] = rootq
            self.depth[rootq] += self.depth[rootp]
            
        else:
            self.id[rootq] = rootp
            self.depth[rootp] += rootq
                
class Grid(object):
    def __init__(self, size, prob):
        self.size = size
        self.prob = prob
        self.elements = UnionFind(self.size)
        self.grid = np.random.choice([1,0], size=(self.size,self.size),
                                     p=[self.prob,1-self.prob])
        self.isOpen = self.grid==1
        
    def __str__(self):
        plt.imshow(self.grid, cmap="gray")
        return 'Grid'

    def _flow(self, isFull, i, j):
        if (i<0) or (i>=self.size):
            return
        if (j<0) or (j>=self.size):
            return
        if not self.isOpen[i][j]:
            return
        if isFull[i][j]:
            return
                
        
        isFull[i][j] = True
        self._flow(isFull, i+1, j)#Down
        self._flow(isFull, i, j+1)#Right
        self._flow(isFull, i, j-1)#Left
        self._flow(isFull, i-1, j)#Up
        

    def flow(self):
        isFull = np.zeros((self.size,self.size))
        for j in range(self.size):
            self._flow(isFull, 0, j)
        return isFull
    
    def percolates(self):
        isFull = self.flow()
        
        n = len(isFull)
        for j in range(n):
            if isFull[n-1][j]:
                return True
        return False
        

G = Grid(10,0.9)
G.percolates()
print(G)    

File: percolation/percolation.cpp
#include <iostream>
#include <stdio.h>
#include <vector>
#include <cstdlib>
#include <tuple>

using namespace std;

vector < vector <int> > genArrays(int size, int prob){
    vector < vector <int> > grid (size, vector <int> (size, 0)); 

    int rn = 0;
    for(int row = 0; row<size; row++){
        for(int col= 0; col<size; col++){
            rn = rand()%100;
            if(rn > prob){
                grid[row][col] = 0;
            }
            else {
                grid[row][col] = 1;
            };
        };
    };
    return grid;
};

void _flow(vector < vector <int> > isFull, vector < vector <int> > isOpen, int i, int j, int size){
    if(i<0 || i>=size) {return;};
    if(j<0 || j>=size) {return;};
    if(isOpen[i][j] == 0) {return;};
    if(isFull[i][j] != 0) {return;};

    isFull[i][j] = 1;
    _flow(isFull, isOpen, i+1, j, size);
    _flow(isFull, isOpen, i, j+1, size);
    _flow(isFull, isOpen, i, j-1, size);
    _flow(isFull, isOpen, i-1, j, size);
    
}

vector < vector <int> > flow(vector < vector <int> > isFull, vector < vector <int> > isOpen, int size){
    for(int j = 0; j<size; j++){
        _flow(isFull, isOpen, 0, j, size);
    };
    return isFull;
};

bool percolates(vector < vector <int> > isOpen, int size) {
    //create isFull array - initialize it with zeros
    vector < vector <int> > isFull (size, vector <int> (size, 0));
    for(int row = 0; row < size; row++){
        for(int col =0; col < size; col++){
            isFull[row][col] = 0;
        };
    };

    isFull = flow(isFull, isOpen, size);
    for(int j=0; j<size; j++){
        if(isFull[size-1][j]==1){
            return 1;
        };
    };
    return 0;

};


int main(){
    int size = 10;
    int prob = 80;
    vector < vector <int> > g = genArrays(size, prob);
    vector < vector <int> > isOpen = g;
    
    //print the grid
    for(int i = 0; i<10; i++){
        for(int j =0; j<10; j++){
            cout<<g[i][j]<<" ";
        };
        cout<<endl;
    };

    int ans = percolates(isOpen, size);
    cout<<ans<<endl;
};